benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java

```diff
@@ -0,0 +1,762 @@
+import com.google.common.base.Stopwatch;
+import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.Random;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+abstract class AbstractBenchmarkRunner implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractBenchmarkRunner.class);
+
+  private final Random random = new Random();
+
+  private final Statistics statistics;
+
+  private final String connectionUrl;
+
+  private final TpccConfiguration tpccConfiguration;
+
+  private final Metrics metrics;
+
+  private boolean failed;
+
+  AbstractBenchmarkRunner(
+      Statistics statistics,
+      String connectionUrl,
+      TpccConfiguration tpccConfiguration,
+      Metrics metrics) {
+    this.statistics = statistics;
+    this.connectionUrl = connectionUrl;
+    this.tpccConfiguration = tpccConfiguration;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void run() {
+    LOG.info("Starting benchmark runner: " + statistics.getRunnerName());
+    try (Connection connection = DriverManager.getConnection(connectionUrl)) {
+      runTransactions(connection);
+      LOG.info("Stopping benchmark runner: " + statistics.getRunnerName());
+    } catch (InterruptedException interruptedException) {
+      LOG.info("Stopping benchmark runner due to interruption: " + statistics.getRunnerName());
+    } catch (Throwable throwable) {
+      throwable.printStackTrace();
+      LOG.error("Benchmark runner failed:" + statistics.getRunnerName(), throwable);
+      failed = true;
+    }
+  }
+
+  private void runTransactions(Connection connection) throws SQLException, InterruptedException {
+    try (Statement statement = connection.createStatement()) {
+      while (true) {
+        try {
+          int transaction = random.nextInt(23);
+          if (transaction < 10) {
+            newOrder(statement);
+            statistics.incNewOrder();
+          } else if (transaction < 20) {
+            payment(statement);
+            statistics.incPayment();
+          } else if (transaction < 21) {
+            orderStatus(statement);
+            statistics.incOrderStatus();
+          } else if (transaction < 22) {
+            delivery(statement);
+            statistics.incDelivery();
+          } else if (transaction < 23) {
+            stockLevel(statement);
+            statistics.incStockLevel();
+          } else {
+            LOG.info("No transaction");
+          }
+          if (Thread.interrupted()) {
+            break;
+          }
+        } catch (Throwable exception) {
+          if ((exception instanceof PSQLException psqlException)
```

Comment:
This will not work when running on the Spanner JDBC driver. Is that intentional?

Reply:
I got this code from https://github.com/GoogleCloudPlatform/pgadapter/blob/1e8b1bf6d5c66c4264c4c780d9bad613e65b371a/benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/BenchmarkRunner.java#L54-L98. I did not change it. Do we have a corresponding exception for the Spanner JDBC driver?

Reply:
This exception is the specific exception that the Spanner JDBC driver will throw if a transaction is aborted: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/e3be3915ecbd73d2c918ee144296152adbd55ef6/spanner/jdbc/src/main/java/com/example/spanner/jdbc/TransactionWithRetryLoopExample.java#L81

More in general, all SQLExceptions that are thrown by the Spanner JDBC driver implement `com.google.cloud.spanner.jdbc.JdbcSqlException`

Reply:
Updated to also handle with `JdbcAbortedException`.